### DISCUSSION TITLE
issue in using `blendpair` between several closed loops

### DISCUSSION THREAD

**Discussion by jimy-byerley** [jimy-byerley/pymadcad#37] [Q&A]
*This is the text from the last question of [tim-rastall-reveal](https://github.com/tim-rastall-reveal)*

Hi again Jimy

Firstly, that hint for using 'Wire' instead of 'wire' worked perfectly and I've now been able to move on to building some geometry.

My next challenge is to create blended surfaces between 2 wires: 

To achieve this, I've been using the 'blendpair' function on 2 wires as follows:

```python
blendpair(wireA,wireB, match='length', tangents='straight')
```
This yields some interesting results where I blend a series of pairs of 12-sided Ngons and 4 sided rectangles as you can see in the image below - the sections are appended into a single mesh that then has end-caps added to form a manifold surface.

I initially thought this issue was a result of the 'handedness' of the vertex order in the wires being inverted but when I checked, everything is right-hand-rule, so in the case of the series of 12 sided Ngons that make up the first 4 sections (from the top) of the mesh shown in the image below, the vertex order and handedness is consistent - with vert index0 on the first wire being directly above vert index0 on the wire below it and so on.

In the case of the 12 sided ngons in the object shown below, the vertices between wires are being linked as follows 0-0, 1-11, 2-10, 3-9, 4-8, 5-7, 6-6 etc

I had assumed that setting the 'match' parameter value to 'closest' would fix this but unfortunately, an error is returned from blending.py as follows: 

>  m = (pts[a] + pts[b] + pts[c] + pts[d]) /4
TypeError: list indices must be integers or slices, not glm.dvec3

Any thoughts? It feels like it could be a bug but equally, I could be miss-understanding something.

I've also attached a csv with the coordinate strings for the object shown in the image below.

I hope this makes sense and thanks for your continued support!

Tim

[AAA-079_inside.csv](https://github.com/jimy-byerley/pymadcad/files/8359464/AAA-079_inside.csv)

![image](https://user-images.githubusercontent.com/82631605/160322843-d293a735-b801-4c33-acff-f29df8c55697.png)

**Comment by jimy-byerley** [comment]
Hello tim ! Sorry for the delay in my answer.

What I see in this figure can be cause by 2 things:

- the "handedness" of the loops as you said (which I suppose are good, looking at your screenshot)
   The winding / handedness of the interfaces passed to `blendpair` must be opposite. Else the point to point matching of the curves will be wrong because the matcher will search each in the wrong direction.
   To solve this problem, you just need to flip the winding of one of the interfaces
   ```python
   blendpair(
      interfaces[i-1], 
      interfaces[i-1].flip(), 
      tangents='straight')
   ```

- the loops phases (which can be a problem in your case because your loops are imported and not generated)
  I suppose you do not know by advance by which point a loop is starting when converted into a `Wire`
  To solve this problem, you can use the `blending.synchronize` function (and writing this I realized it was not in the docs yet)
  ```python
  blendpair(
       interfaces[i-1], 
       blending.synchronize(interfaces[i-1], interfaces[i]).flip(), 
       tangents='straight')
  ```

This is what allows me to get this result:
```python
profiles = [
	regon((O,Z), 1, 12),
	regon((-0.5*Z,Z), 1, 4) .transform(angleAxis(1,Z)),   # rotate that profile, so it start is no longer face to the other profile's starts
	regon((-1*Z,Z), 1.2, 12),
	regon((-1.5*Z,Z), 1.2, 4),
	]

result = reduce(add, (
	blendpair(
		profiles[i-1], 
		blending.synchronize(
			profiles[i-1], 
			profiles[i]).flip(),
		tangents='straight')
	for i in range(1,len(profiles))
	))
```
![image](https://user-images.githubusercontent.com/7940753/161447566-02088b02-ee28-4a63-8caf-e5003d6bd793.png)
However the result is not that great, because my current implementation of the matching procedure is not great. The patching procedure is what you select with `match='length'` or `match='closest'`. `'closest'` is not working at the moment because I started to improve things some times ago and still didn't finished.
You can notice that the first blending at the top is quite relevant, but the second one is a twisted despite `synchronize`, and the thirst is twisted as well. All this because the current matcher makes mistakes.

So this might work better in coming weeks but I do not have a better solution right now ...

**Reply by tim-rastall-reveal** [reply]
Hey Jimy

Thanks for the explanation - I'm pleased that my assessment of the match='closest' parameter not working as intended was correct :D.
I can easily flip every other section in my script and now I know what the current limitations are, I will add some pre-processing steps to ensure that the first index of every section is the one with the highest 'y' value so that the starting point for each wire is as close as it can be to the next one. I'll also add a couple of equally-spaced vertices between the 4 verts of the rectangular sections so the vertices count for all sections is the same.

I'll look forward to seeing updates to the blending module when you have time! It may prove to be quite popular as I couldn't find many python libraries that performed loft operations well :).

Cheers

Tim

**Comment by tim-rastall-reveal** [comment]
Hey Jimy

Ok, just tested my approach and it works well! One last question - is there a way to adjust the number of sub-loops the blendpair function adds between loops?  I'm happy with a blocky/tessellated look so those extra triangles aren't necessary for my purposes and push the overall tri-count up. Is it the 'resolution' parameter that controls this? I couldn't find anything in the documentation about it.

Many thanks again!

Tim

**Reply by jimy-byerley** [reply]
Hey again
There is the `resolution` parameter described in [madcad.primitives](https://pymadcad.readthedocs.io/en/latest/reference/primitives.html?highlight=resolution#curve-resolution)
In madcad, all functions with a `resolution` parameter support those specifications. So you have 4 different ways to give the resolution, depending on whether you want it to be function of the part scale, curvature, or fixed.
It seems also that the list needs to be updated: `'fixed'` has been renamed `'div'` long ago.

If you want for instance exactly 2 intermediate profiles to be generated, you can set `resolution=('div', 2)`
